apps/hash-frontend/src/blocks/page/create-suggester/mention-suggester.tsx

```diff
@@ -1,188 +1,632 @@
+import { MentionSuggesterSubheading } from "./mention-suggester/mention-suggester-subheading";
+import { MentionSuggesterWrapper } from "./mention-suggester/mention-suggester-wrapper";
+
+export type Mention =
+  | {
+      kind: "entity";
+      entityId: EntityId;
+    }
+  | {
+      kind: "user";
+      entityId: EntityId;
+    }
+  | {
+      kind: "page";
+      entityId: EntityId;
+    }
+  | {
+      kind: "property-value";
+      entityId: EntityId;
+      propertyTypeId: VersionedUrl;
+    }
+  | {
+      kind: "outgoing-link";
+      entityId: EntityId;
+      linkEntityId: EntityId;
+    };
+
+export type MentionKind = Mention["kind"];
 
-export type MentionType = "user" | "page" | "entity";
 export interface MentionSuggesterProps {
   search?: string;
-  onChange(entityId: EntityId, mentionType: MentionType): void;
+  onChange(mention: Mention): void;
   ownedById: OwnedById;
 }
 
-type SearchableItem = {
-  icon?: string | null;
-  shortname?: string;
-  name: string;
-  desc?: string;
-  entityId: EntityId;
-  mentionType: MentionType;
-  isActiveOrgMember?: boolean;
-};
+type EntitiesByType = {
+  entityType: EntityTypeWithMetadata;
+  displayedEntities: Entity[];
+  allEntities: Entity[];
+}[];
+
+const numberOfEntitiesDisplayedPerSection = 4;
 
 export const MentionSuggester: FunctionComponent<MentionSuggesterProps> = ({
   search = "",
   onChange,
-  ownedById,
+  ownedById: _ownedById,
 }) => {
-  const { users, loading: usersLoading } = useUsers();
-  const { entitiesSubgraph, loading: entitiesLoading } = useQueryEntities({
-    excludeEntityTypeIds: [
-      types.entityType.user.entityTypeId,
-      types.entityType.page.entityTypeId,
-    ],
+  const { authenticatedUser } = useAuthenticatedUser();
+  const wrapperRef = useRef<HTMLDivElement>(null);
+
+  const { propertyTypes } = usePropertyTypes();
+
+  const [selectedEntityIndex, setSelectedEntityIndex] = useState(0);
+
+  const [expandedEntityTypes, setExpandedEntityTypes] = useState<
+    VersionedUrl[]
+  >([]);
+
+  const selectedEntityRef = useRef<HTMLDivElement>(null);
+
+  const [displayEntitySubMenu, setDisplayEntitySubMenu] = useState(false);
+
+  const [entitySubMenuSortOrder, setEntitySubMenuSortOrder] =
+    useState<SortOrder>("asc");
+
+  useScrollLock(displayEntitySubMenu, wrapperRef.current ?? undefined);
+
+  const [entitySelectedSubMenuIndex, setEntitySelectedSubMenuIndex] =
+    useState(0);
+
+  // scroll the selected option into view
+  useEffect(
+    () => selectedEntityRef.current?.scrollIntoView({ block: "nearest" }),
+    [selectedEntityIndex],
+  );
+
+  const { data, loading: loadingEntities } = useQuery<
+    StructuralQueryEntitiesQuery,
+    StructuralQueryEntitiesQueryVariables
+  >(structuralQueryEntitiesQuery, {
+    variables: {
+      query: {
+        filter: {
+          any: [
+            {
+              equal: [
+                { path: ["ownedById"] },
+                { parameter: authenticatedUser.accountId },
```

Comment:
Re. your question in a comment, I think we should include entities owned by orgs the user is a member of.

Reply:
I've included org entities in https://github.com/hashintel/hash/pull/3244/commits/64e2aa4b5767f87b8fc08cd35675224d273792d6, have also ensured that the user's entities are displayed before org entities in each section (a suggestion from @nonparibus)

Reply:
We do want to include **all entities**, not just those a user is a member of. Although we can wait until permissions/authorization are in until we introduce this.

Stuff from a user's own webs (their own, plus those they are a member of) ought to be heavily weighted to generally appear above external ones in the list (exceptions may be made for exact name match things from outside a graph, and/or which are extremely popular and widely used, or which are system types, etc... pending full search algo determination). @CiaranMn @benwerner01